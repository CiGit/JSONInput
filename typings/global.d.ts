declare module 'baobab-react/higher-order' {
    export const root: any;
    export const branch: any;
}
declare class Baobab {
    constructor(...args: {}[]);
}
declare module "baobab" {
    export default Baobab;
}

declare module 'react-pure-render/function' {
    export default function (args: {}[]): boolean
}
declare module 'jsonschema' {
    export class Validator {
        public attributes: { [key: string]: (...args: any[]) => any }
        public validate: (value: {} | void, schema: {}, options: { [option: string]: {} }) => any
    }
    export class SchemaError {
        constructor(message: string);
    }
}
declare module 'prop-types' {
    export default React.PropTypes;
}