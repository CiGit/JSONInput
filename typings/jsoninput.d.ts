import * as React from 'react';
import { Schema } from './types';
import { ValidationError } from 'jsonschema/lib';

type WidgetMap = {
    [key: string]: React.ComponentType<any>;
};
export function setDefaultWidgets(map: WidgetMap): void;

declare class Form extends React.Component<
    {
        schema?: {};
        value?: {};
        onChange: (value: any, errors: ValidationError[]) => void;
    },
    any
> {
    /**
     * Retrieve current form's value
     */
    getValue: () => {} | void;
    /**
     * Validate the entire form and pass errors down to views
     *
     */
    validate: () => ValidationError[];
}
export { Schema };
export default Form;
