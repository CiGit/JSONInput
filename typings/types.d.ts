import React from 'react';
import { Schema as JsonSchema } from "jsonschema/lib";

type ErrorFn = (value: {}, formValue: {}) => string

type TYPESTRING =
    | 'string'
    | 'number'
    | 'boolean'
    | 'object'
    | 'array'
    | 'null';

type View = {
    type?:
    | string
    | React.ComponentClass<WidgetProps>
    | React.SFC<WidgetProps>,
    [key: string]: any
};
declare namespace Schema {
    type BASE = {
        type: TYPESTRING | TYPESTRING[];
        value?: {};
        visible?: (value?: {}, formValue?: {}) => boolean;
        errored?: ErrorFn;
        index?: number;
        view?: View;
        required?: boolean;
    }
    type String = BASE & {
        type: 'string' | ['string', 'null'];
        maxLength?: number;
        minLength?: number;
    }
    type Number = BASE & {
        type: 'number' | ['number', 'null'];
        maximum?: number;
        minimum?: number
    }
    type Boolean = BASE & {
        type: 'boolean' | ['boolean', 'null'];
    }
    type Object = BASE & {
        type: 'object' | ['object', 'null'];
        properties?: { [property: string]: Schema };
        additionalProperties?: Schema;
    }
    type Array = BASE & {
        type: 'array' | ['array', 'null']
        items?: Schema | Schema[];
        additionalItems?: Schema;
        maxItems?: number;
        minItems?: number;
    }
}
type Schema = Schema.BASE | Schema.String | Schema.Number | Schema.Boolean | Schema.Array | Schema.Object;
declare namespace WidgetProps {
    /**
     * Use for string / number / boolean widgets
     */
    type BaseProps = {
        value: {},
        onChange: (value: {}) => void,
        schema: Schema,
        view: View,
        errorMessage?: string[],
        editKey: string,
        path: string[]
    }
    /**
     * Use for Object widget
     */
    type ObjectProps = BaseProps & {
        children?: (React.ComponentClass<WidgetProps> | React.SFC<WidgetProps>)[],
        addKey: (key: string, value?: {}) => void,
        removeKey: (key: string) => void,
        alterKey: (key: string, newKey: string) => void,
    }
    /**
     * Use for Array widget
     */
    type ArrayProps = BaseProps & {
        children?: (React.ComponentClass<WidgetProps> | React.SFC<WidgetProps>)[],
        onChildAdd: () => void,
        onChildRemove: (index: number) => void
    };
}
    /**
     * Props passed in widgets.
     */
    type WidgetProps = WidgetProps.ArrayProps | WidgetProps.BaseProps | WidgetProps.ObjectProps

type Action = (tree: any, path?: string[], ...args: {}[]) => {} | void;
