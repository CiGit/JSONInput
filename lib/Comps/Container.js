'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Store = require('../Store');

var _Store2 = _interopRequireDefault(_Store);

var _higherOrder = require('baobab-react/higher-order');

var _SchemaType = require('./SchemaType');

var _SchemaType2 = _interopRequireDefault(_SchemaType);

var _actions = require('../Store/actions');

var actions = _interopRequireWildcard(_actions);

var _customValidator = require('./../Utils/customValidator');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var BranchedSchemaType = (0, _higherOrder.branch)(_SchemaType2.default, {
    cursors: {
        schema: 'schema',
        status: 'status',
        value: 'value'
    }
});

/**
 * Top Component
 */

var Container = function (_React$Component) {
    _inherits(Container, _React$Component);

    function Container(props) {
        _classCallCheck(this, Container);

        var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Container).call(this, props));

        _this.tree = (0, _Store2.default)();
        _this.tree.select('value').set(props.value);
        _this.tree.select('schema').set(props.schema);
        _this.tree.select('value').on('update', function (event) {
            return _this.props.onChange(event.data.currentData);
        });

        // baobab-react optim as actions are bound on each render -> pure
        _this.ACTIONS = {};
        Object.keys(actions).forEach(function (action) {
            _this.ACTIONS[action] = actions[action].bind(_this.tree, _this.tree);
            return;
        });
        return _this;
    }

    _createClass(Container, [{
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
            this.tree.select('value').set(nextProps.value);
            if (this.props.schema !== nextProps.schema) {
                this.tree.select('schema').set(nextProps.schema);
            }
        }
    }, {
        key: 'shouldComponentUpdate',
        value: function shouldComponentUpdate(nextProps) {
            return nextProps.value !== this.tree.get('value') || nextProps.schema !== this.props.schema;
        }
    }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
            this.tree.release();
        }
    }, {
        key: 'getValue',
        value: function getValue() {
            return this.tree.get('value');
        }
    }, {
        key: 'validate',
        value: function validate() {
            var validationResult = (0, _customValidator.validate)(this.tree.get('value'), this.tree.get('schema'), this.ACTIONS.getFormValue());
            var setErrors = this.ACTIONS.setErrors;

            var errorMap = new Map();
            // Collect each error associated with a given path
            validationResult.errors.forEach(function (error) {
                var errors = errorMap.get(error.property) || [];
                errors.push(error.message); // Add new error
                errorMap.set(error.property, errors);
            });
            errorMap.forEach(function (value, key) {
                setErrors(key.split(/\.|\[|\]/).filter(function (x) {
                    return x !== '';
                }).slice(1), value);
            });
            return validationResult.errors;
        }
    }, {
        key: 'render',
        value: function render() {
            var Rooted = (0, _higherOrder.root)(BranchedSchemaType, this.tree);
            return _react2.default.createElement(Rooted, _extends({}, this.props, {
                path: [],
                actions: this.ACTIONS
            }));
        }
    }]);

    return Container;
}(_react2.default.Component);

exports.default = Container;