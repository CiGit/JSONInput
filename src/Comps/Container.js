import React, { PropTypes } from 'react';
import createTree from '../Store';
import { root, branch } from 'baobab-react/higher-order';
import SchemaType from './SchemaType';
import * as actions from '../Store/actions';
import { validate } from './../Utils/customValidator';

const BranchedSchemaType = branch(SchemaType, {
    cursors: {
        schema: 'schema',
        status: 'status',
        value: 'value'
    }
});

/**
 * Top Component
 */
class Container extends React.Component {
    constructor(props) {
        super(props);
        this.tree = createTree();
        this.tree.select('value').set(props.value);
        this.tree.select('schema').set(props.schema);
        this.tree.select('value')
            .on('update', event => this.props.onChange(event.data.currentData));

        // baobab-react optim as actions are bound on each render -> pure
        this.ACTIONS = {};
        Object.keys(actions)
            .forEach(action => {
                this.ACTIONS[action] = actions[action].bind(this.tree, this.tree);
                return;
            });
    }
    componentWillReceiveProps(nextProps) {
        this.tree.select('value').set(nextProps.value);
        if (this.props.schema !== nextProps.schema) {
            this.tree.select('schema').set(nextProps.schema);
        }
    }
    shouldComponentUpdate(nextProps) {
        return nextProps.value !== this.tree.get('value') || nextProps.schema !== this.props.schema;
    }
    componentWillUnmount() {
        this.tree.release();
    }
    getValue() {
        return this.tree.get('value');
    }
    validate() {
        const validationResult = validate(this.tree.get('value'),
            this.tree.get('schema'), this.ACTIONS.getFormValue());
        const { setErrors } = this.ACTIONS;
        const errorMap = new Map();
        // Collect each error associated with a given path
        validationResult.errors.forEach(error => {
            const errors = errorMap.get(error.property) || [];
            errors.push(error.message); // Add new error
            errorMap.set(error.property, errors);
        });
        errorMap.forEach((value, key) => {
            setErrors(key.split(/\.|\[|\]/).filter(x => x !== '').slice(1), value);
        });
        return validationResult.errors;
    }
    render() {
        const Rooted = root(BranchedSchemaType, this.tree);
        return (
            <Rooted
                {...this.props}
                path={[]}
                actions={this.ACTIONS}
            />);
    }
}
Container.propTypes = {
    onChange: PropTypes.func,
    schema: PropTypes.object,
    value: PropTypes.any
};

export default Container;
