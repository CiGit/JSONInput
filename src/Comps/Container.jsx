import React, { PropTypes } from 'react';
import createTree from '../Store';
import { root, branch } from 'baobab-react/higher-order';
import SchemaType from './SchemaType.jsx';
import * as actions from '../Store/actions';
import { validate } from './../Utils/customValidator';

const BranchedSchemaType = branch(SchemaType, {
    cursors: {
        schema: 'schema',
        status: 'status',
        value: 'value'
    } /* ,
    actions: {
        update,
        defaultValue,
        getStatus,
        getErrors
    }*/
});

/**
 * Top Component
 */
class Container extends React.Component {
    constructor(props) {
        super(props);
        this.tree = createTree();
        this.tree.select('schema').set(props.schema);
        this.tree.select('value').set(props.value);
        this.tree.select('value')
            .on('update', event => this.props.onChange(event.data.currentData));

        // baobab-react optim as actions are bound on each render -> pure
        this.ACTIONS = {};
        Object.keys(actions)
            .forEach(action => this.ACTIONS[action] = actions[action].bind(this.tree, this.tree));
    }
    componentWillReceiveProps(nextProps) {
        this.tree.select('schema').set(nextProps.schema);
        this.tree.select('value').set(nextProps.value);
    }
    componentWillUnmount() {
        this.tree.release();
    }
    validate() {
        const validationResult = validate(this.tree.get('value'), this.props.schema);
        const { setErrors } = this.ACTIONS;
        const errorMap = new Map();
        // Collect each error associated with a given path
        validationResult.errors.forEach(error => {
            const errors = errorMap.get(error.property) || [];
            errors.push(error.message); // Add new error
            errorMap.set(error.property, errors);
        });
        errorMap.forEach((value, key) => {
            setErrors(key.split('.').slice(1), value);
        });
        return validationResult.errors;
    }
    render() {
        const Rooted = root(BranchedSchemaType, this.tree);
        return (<Rooted path={ [] }
                        actions={ this.ACTIONS } />);
    }
}
Container.propTypes = {
    onChange: PropTypes.func,
    schema: PropTypes.object,
    value: PropTypes.any
};

export default Container;
