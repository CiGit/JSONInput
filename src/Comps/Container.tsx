import * as React from 'react';
import { root, branch } from 'baobab-react/higher-order';
import createTree from '../Store/index';
import SchemaType from './SchemaType';
import { setErrors } from '../Store/actions';
import validate from './../Utils/customValidator';

import { Schema } from '../../typings/types';

const BranchedSchemaType = branch(
    {
        schema: 'schema',
        status: 'status',
        value: 'value'
    },
    SchemaType
);

export type Props = {
    onChange: (value: {}, errors: {}[]) => void,
    schema: Schema,
    value?: {}
};
function noop() { };
/**
 * Top Component
 */
class Container extends React.Component<Props, undefined> {
    event: boolean;
    static defaultProps = { schema: {} };
    private tree: any;
    private rooted: React.StatelessComponent<{
        onChange: (value: {}, errors?: {}[]) => void,
        path: string[]
    }>;
    props: Props;

    constructor(props: Props) {
        super(props);
        this.tree = createTree();
        this.updateTree(props.value, props.schema);
        this.rooted = root(this.tree, BranchedSchemaType);
    }
    componentDidMount() {
        this.tree
            .select('value')
            .on('update', (event: { data: { currentData: {} } }) => {
                if (this.event) {
                    this.props.onChange(
                        event.data.currentData,
                        validate(
                            event.data.currentData,
                            this.tree.get('schema'),
                            event.data.currentData
                        ).errors
                    )
                }
            }
            );
    }

    componentWillReceiveProps(nextProps: Props) {
        if (
            nextProps.value === this.tree.get('value') &&
            nextProps.schema === this.props.schema
        ) {
            return;
        }
        this.updateTree(nextProps.value, nextProps.schema);
    }
    componentWillUnmount() {
        this.tree.release();
    }
    shouldComponentUpdate() {
        return false;
    }
    getValue() {
        return this.tree.get('value');
    }
    updateTree(value?: {}, schema?: Schema) {
        this.event = false;
        this.tree.set('value', value);
        this.tree.set('schema', schema);
        this.tree.set('status', {});
        this.tree.commit();
        this.event = true;
    }
    validate() {
        const validationResult = validate(
            this.tree.get('value'),
            this.tree.get('schema'),
            this.tree.get('value')
        );
        const errorMap = new Map<string, string[]>();
        // Collect each error associated with a given path
        validationResult.errors.forEach(error => {
            const errors = errorMap.get(error.property) || [];
            errors.push(error.message); // Add new error
            errorMap.set(error.property, errors);
        });
        errorMap.forEach((value, key) => {
            setErrors(
                this.tree,
                key.split(/\.|\[|\]/).filter(x => x !== '').slice(1),
                value
            );
        });
        return validationResult.errors;
    }
    render() {
        const Rooted = this.rooted;
        return <Rooted onChange={noop} path={[]} />;
    }
}

export default Container;
