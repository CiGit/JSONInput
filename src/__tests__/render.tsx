import * as React from 'react';
import { render, cleanup } from 'react-testing-library';
import Container, { setDefaultWidgets } from '../index';
import defaultViews from '../Comps/Views';
import { Schema } from '../../typings/jsoninput';

it('renders a fallback view', () => {
  const { container, unmount } = render(
    <Container schema={{ type: 'number' }} onChange={() => {}} />,
  );
  expect(container.firstChild).toMatchSnapshot();
  unmount();
});
describe('Render fields with default values', () => {
  beforeAll(() => {
    setDefaultWidgets(defaultViews as any);
  });
  afterEach(cleanup);
  function renderDefault(schema: Schema) {
    return new Promise(resolve => {
      const { container, unmount } = render(
        <Container schema={schema} onChange={resolve} />,
      );
      expect(container.firstChild).toMatchSnapshot();
      unmount();
    }).then(v => {
      expect(v).toEqual(schema.value);
      return v;
    });
  }
  test('Default string', () => {
    return renderDefault({ type: 'string', value: 'String value' });
  });
  test('Default number', () => {
    return renderDefault({ type: 'number', value: 4 });
  });
  test('Default boolean', () => {
    return renderDefault({ type: 'boolean', value: false });
  });
  test('Default object', () => {
    const schema: Schema = { type: 'object', value: { a: 1 } };
    return renderDefault(schema).then(v => expect(v).not.toBe(schema.value));
  });
  test('Default array', () => {
    const schema: Schema = { type: 'array', value: ['string'] };
    return renderDefault(schema).then(v => expect(v).not.toBe(schema.value));
  });
  test('Default tuple', () => {
    const schema: Schema = {
      type: 'array',
      value: [undefined, undefined],
      items: [
        { type: 'string', value: 'Hello' },
        { type: 'number', value: 42 },
      ],
    };
    return new Promise(resolve => {
      render(<Container schema={schema} onChange={resolve} />);
    }).then(v => expect(v).toEqual(['Hello', 42]));
  });
  test('Composite values', () => {
    return new Promise(resolve => {
      render(
        <Container
          schema={{
            type: 'object',
            properties: {
              arrayProp: {
                type: 'array',
                value: [{}],
                items: {
                  type: 'object',
                  properties: {
                    ap: {
                      type: 'string',
                      value: 'key string',
                    },
                  },
                },
              },
              numberProps: {
                value: 1,
              },
            },
          }}
          onChange={resolve}
        />,
      );
    }).then(v =>
      expect(v).toEqual({
        arrayProp: [{ ap: 'key string' }],
        numberProps: 1,
      }),
    );
  });
});

describe('Handle the unknown', () => {
  beforeAll(() => {
    setDefaultWidgets(defaultViews as any);
  });
  test('Null', () => {
    const { container } = render(
      <Container
        schema={{
          type: ['null'],
        }}
        onChange={() => {}}
      />,
    );
    expect(container.firstChild).toMatchSnapshot('Undefined view null');
  });
  test('Unknown', () => {
    const { container } = render(
      <Container
        schema={{
          type: 'unknown' as any,
        }}
        onChange={() => {}}
      />,
    );
    expect(container.firstChild).toMatchSnapshot('Undefined view unknow');
  });
  test('Inference', () => {
    const container = document.createElement('div');
    const schema = {};
    render(
      <Container
        schema={schema}
        value={{
          number: 1,
          string: '',
          boolean: false,
          array: [],
          object: {},
        }}
        onChange={() => {}}
      />,
      { container, baseElement: container },
    );
    expect(container.firstChild).toMatchSnapshot();
    let form: Container;
    render(
      <Container
        schema={schema}
        ref={f => (form = f!)}
        value={{
          // *Wrong* types
          number: '1',
          string: 3,
          boolean: 'false',
          array: {},
          object: [],
        }}
        onChange={() => {}}
      />,
      { container, baseElement: container },
    );
    // No errors after new value
    expect(form!.validate()).toHaveLength(0);
  });
  test("Doesn't send update if value didn't change", () => {
    const schema = {};
    const container = document.createElement('div');
    const onChange = jest.fn();
    for (let i = 0; i < 10; i++) {
      render(<Container schema={schema} value={i} onChange={onChange} />, {
        container,
        baseElement: container,
      });
    }
    expect(onChange).not.toBeCalled();
  });
  test("Doesn't send update for same input", () => {
    const schema = {};
    const value = { a: 1 };
    const container = document.createElement('div');
    const onChange = jest.fn();
    for (let i = 0; i < 10; i++) {
      render(<Container schema={schema} value={value} onChange={onChange} />, {
        container,
        baseElement: container,
      });
    }
    expect(onChange).not.toBeCalled();
  });
  test('Render a specified view', () => {
    const { container } = render(
      <Container
        schema={{ type: 'number', view: { type: 'arrowNumber' } }}
        onChange={() => {}}
      />,
    );
    expect(container.querySelectorAll('[type="number"]').length).toBe(1);
  });
  test('Render a component', () => {
    const text = "I'm a dummy component";
    const { getByText } = render(
      <Container
        schema={{
          type: 'string',
          view: {
            type: () => {
              return <div>{text}</div>;
            },
          },
        }}
        onChange={() => {}}
      />,
    );
    expect(() => getByText(text)).not.toThrow();
  });
  test('Render a multiple type', () => {
    const { container } = render(
      <Container
        schema={{
          type: ['null', 'string', 'number'],
        }}
        onChange={() => {}}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
