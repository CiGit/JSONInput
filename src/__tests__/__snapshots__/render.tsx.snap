// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handle the unknown Inference 1`] = `
<div>
  <label
    class="objectField "
  >
    <span
      class="title"
    />
    <div>
      <div>
        <label
          class="numberField "
        >
          <span
            class="title"
          >
            number
          </span>
          <input
            type="string"
            value="1"
          />
          <span />
          <span />
        </label>
      </div>
      <div>
        <label
          class="stringField "
        >
          <span
            class="title"
          >
            string
          </span>
          <input
            type="string"
            value=""
          />
          <span />
          <span />
        </label>
      </div>
      <div>
        <label
          class="booleanField "
        >
          <span
            class="title"
          >
            boolean
          </span>
          <input
            type="checkbox"
            value="false"
          />
          <span />
          <span />
        </label>
      </div>
      <div>
        <label
          class="arrayField "
        >
          <span
            class="title"
          >
            array
          </span>
          <div>
            <div />
            <button>
              +
            </button>
          </div>
          <span />
          <span />
        </label>
      </div>
      <div>
        <label
          class="objectField "
        >
          <span
            class="title"
          >
            object
          </span>
          <div />
          <span />
          <span />
        </label>
      </div>
    </div>
    <span />
    <span />
  </label>
</div>
`;

exports[`Handle the unknown Null: Undefined view null 1`] = `
<div>
  Undefined field type "null", []
</div>
`;

exports[`Handle the unknown Render a multiple type 1`] = `
<div>
  <label
    class="null,string,numberField "
  >
    <span
      class="title"
    />
    <input
      type="string"
      value=""
    />
    <span />
    <span />
  </label>
</div>
`;

exports[`Handle the unknown Unknown: Undefined view unknow 1`] = `
<div>
  Undefined field type "unknown", []
</div>
`;

exports[`Render fields with default values Default array 1`] = `
<div>
  <label
    class="arrayField "
  >
    <span
      class="title"
    />
    <div>
      <div>
        <div>
          <button>
            -
          </button>
          <div>
            <label
              class="stringField "
            >
              <span
                class="title"
              >
                0
              </span>
              <input
                type="string"
                value="string"
              />
              <span />
              <span />
            </label>
          </div>
        </div>
      </div>
      <button>
        +
      </button>
    </div>
    <span />
    <span />
  </label>
</div>
`;

exports[`Render fields with default values Default boolean 1`] = `
<div>
  <label
    class="booleanField "
  >
    <span
      class="title"
    />
    <input
      type="checkbox"
      value="false"
    />
    <span />
    <span />
  </label>
</div>
`;

exports[`Render fields with default values Default number 1`] = `
<div>
  <label
    class="numberField "
  >
    <span
      class="title"
    />
    <input
      type="string"
      value="4"
    />
    <span />
    <span />
  </label>
</div>
`;

exports[`Render fields with default values Default object 1`] = `
<div>
  <label
    class="objectField "
  >
    <span
      class="title"
    />
    <div>
      <div>
        <label
          class="numberField "
        >
          <span
            class="title"
          >
            a
          </span>
          <input
            type="string"
            value="1"
          />
          <span />
          <span />
        </label>
      </div>
    </div>
    <span />
    <span />
  </label>
</div>
`;

exports[`Render fields with default values Default string 1`] = `
<div>
  <label
    class="stringField "
  >
    <span
      class="title"
    />
    <input
      type="string"
      value="String value"
    />
    <span />
    <span />
  </label>
</div>
`;

exports[`renders a fallback view 1`] = `
<span>
  Widget for 'number' was not defined
</span>
`;
